sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamEnumerableConverter.java

```diff
@@ -111,6 +141,43 @@ private static PipelineResult run(
     return result;
   }
 
+  private static PipelineResult limitRun(
+      PipelineOptions options,
+      BeamRelNode node,
+      DoFn<Row, Void> doFn,
+      LimitStateWrapper stateWrapper) {
+    ExecutorService pool = Executors.newFixedThreadPool(1);
+
+    options.as(DirectOptions.class).setBlockOnRun(false);
+    Pipeline pipeline = Pipeline.create(options);
+    PCollectionTuple.empty(pipeline).apply(node.toPTransform()).apply(ParDo.of(doFn));
+
+    PipelineResult result = pipeline.run();
+
+    pool.execute(
+        new Runnable() {
+          @Override
+          public void run() {
+            while (!result.getState().isTerminal()) {
+              try {
+                Thread.sleep(DaemonThreadSleepIntervalMillis);
```

Comment:
I think result.waitUntilFinish(duration) is good here: https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/PipelineResult.java#L58

Reply:
Great suggestion!